File: go/acceptance_test.go
// ABOUTME: Acceptance tests for the promotion engine system.
// ABOUTME: Tests end-to-end behavior of pricing carts with promotions.
package promotionengine_test

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	pe "promotionengine"
)

// TestPromotionEngineWithNoPromotions verifies that the engine calculates
// correct prices for a cart with no promotions applied.
func TestPromotionEngineWithNoPromotions(t *testing.T) {
	// Create products
	apple := pe.Product{
		Sku:      pe.Sku{Value: "APPLE"},
		Name:     "Apple",
		Category: "Fruit",
	}
	banana := pe.Product{
		Sku:      pe.Sku{Value: "BANANA"},
		Name:     "Banana",
		Category: "Fruit",
	}

	// Create cart with line items
	cart := pe.Cart{
		Lines: []pe.LineItem{
			{
				Product:   apple,
				Quantity:  pe.Quantity{Value: 3},
				UnitPrice: pe.Euros(1.50),
			},
			{
				Product:   banana,
				Quantity:  pe.Quantity{Value: 2},
				UnitPrice: pe.Euros(0.75),
			},
		},
	}

	// Create pricing context
	now := time.Date(2025, 12, 18, 10, 0, 0, 0, time.UTC)
	context := pe.PricingContext{
		Now:          now,
		Channel:      "online",
		CustomerID:   "CUST123",
		CustomerTags: map[string]bool{"new": true},
	}

	// Create engine with no promotions
	engine := pe.NewPromotionEngine([]pe.Promotion{})

	// Price the cart
	summary := engine.Price(cart, context)

	// Verify results
	expectedSubtotal := pe.Euros(1.50).Multiply(3).Add(pe.Euros(0.75).Multiply(2))
	assert.Equal(t, expectedSubtotal, summary.Subtotal, "Subtotal should be sum of line totals")
	assert.Equal(t, pe.CentsAmount(0), summary.DiscountTotal, "No discounts should be applied")
	assert.Equal(t, expectedSubtotal, summary.Total, "Total should equal subtotal with no discounts")
	assert.Empty(t, summary.AppliedDiscounts, "No discounts should be in the list")
}

// TestPromotionEngineWithStubPromotion verifies that the engine correctly
// applies a simple stub promotion to demonstrate the promotion mechanism.
func TestPromotionEngineWithStubPromotion(t *testing.T) {
	// Create products
	apple := pe.Product{
		Sku:      pe.Sku{Value: "APPLE"},
		Name:     "Apple",
		Category: "Fruit",
	}

	// Create cart
	cart := pe.Cart{
		Lines: []pe.LineItem{
			{
				Product:   apple,
				Quantity:  pe.Quantity{Value: 3},
				UnitPrice: pe.Euros(1.50),
			},
		},
	}

	// Create pricing context
	now := time.Date(2025, 12, 18, 10, 0, 0, 0, time.UTC)
	context := pe.PricingContext{
		Now:          now,
		Channel:      "online",
		CustomerID:   "CUST123",
		CustomerTags: map[string]bool{"new": true},
	}

	// Create a stub promotion that gives 10% off
	stubPromo := &pe.StubPromotion{
		PromoID:          pe.PromotionID{Value: "STUB10"},
		DiscountAmount:   pe.Euros(0.45), // 10% of 4.50
		ShouldApply:      true,
		Target:           "cart",
		DiscountDetails:  "Stub 10% discount",
	}

	// Create engine with stub promotion
	engine := pe.NewPromotionEngine([]pe.Promotion{stubPromo})

	// Price the cart
	summary := engine.Price(cart, context)

	// Verify results
	expectedSubtotal := pe.Euros(4.50)
	expectedDiscount := pe.Euros(0.45)
	expectedTotal := pe.Euros(4.05)

	assert.Equal(t, expectedSubtotal, summary.Subtotal, "Subtotal should be 4.50")
	assert.Equal(t, expectedDiscount, summary.DiscountTotal, "Discount should be 0.45")
	assert.Equal(t, expectedTotal, summary.Total, "Total should be 4.05")
	require.Len(t, summary.AppliedDiscounts, 1, "One discount should be applied")

	discount := summary.AppliedDiscounts[0]
	assert.Equal(t, pe.PromotionID{Value: "STUB10"}, discount.PromotionID)
	assert.Equal(t, expectedDiscount, discount.Amount)
	assert.Equal(t, "cart", discount.Target)
	assert.Equal(t, "Stub 10% discount", discount.Details)
}

File: java/src/test/java/com/promoengine/engine/AcceptanceTest.java
// ABOUTME: Acceptance tests for the promotion engine system.
// ABOUTME: Tests end-to-end behavior of pricing carts with promotions.
package com.promoengine.engine;

import com.promoengine.domain.*;
import com.promoengine.pricing.*;
import com.promoengine.promotions.StubPromotion;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static org.assertj.core.api.Assertions.*;

class AcceptanceTest {

    private PricingContext context;
    private Product productA;
    private Product productB;
    private Product productC;

    @BeforeEach
    void setUp() {
        context = new PricingContext(
            Instant.parse("2025-12-18T10:00:00Z"),
            "online",
            "CUST001",
            Set.of("regular")
        );

        productA = new Product(
            new Sku("SKU-A"),
            "Product A",
            "electronics"
        );

        productB = new Product(
            new Sku("SKU-B"),
            "Product B",
            "electronics"
        );

        productC = new Product(
            new Sku("SKU-C"),
            "Product C",
            "books"
        );
    }

    @Test
    void shouldCalculateSubtotalWithoutPromotions() {
        Cart cart = new Cart(List.of(
            new LineItem(productA, new Quantity(2), Money.euros("10.00")),
            new LineItem(productB, new Quantity(1), Money.euros("25.00"))
        ));

        PromotionEngine engine = new PromotionEngine(List.of());
        PriceSummary summary = engine.price(cart, context);

        assertThat(summary.subtotal()).isEqualTo(Money.euros("45.00"));
        assertThat(summary.discountTotal()).isEqualTo(Money.euros("0.00"));
        assertThat(summary.total()).isEqualTo(Money.euros("45.00"));
        assertThat(summary.appliedDiscounts()).isEmpty();
    }

    @Test
    void shouldApplySinglePromotion() {
        Cart cart = new Cart(List.of(
            new LineItem(productA, new Quantity(2), Money.euros("10.00"))
        ));

        AppliedDiscount discount = new AppliedDiscount(
            new PromotionId("PROMO-10-OFF-A"),
            Money.euros("2.00"),
            "line",
            "$2 off SKU-A",
            Map.of(new Sku("SKU-A"), Money.euros("2.00"))
        );

        StubPromotion promotion = new StubPromotion(
            new PromotionId("PROMO-10-OFF-A"),
            true,
            List.of(discount)
        );

        PromotionEngine engine = new PromotionEngine(List.of(promotion));
        PriceSummary summary = engine.price(cart, context);

        // 2 * $10.00 = $20.00, $2 off = $18.00
        assertThat(summary.subtotal()).isEqualTo(Money.euros("20.00"));
        assertThat(summary.discountTotal()).isEqualTo(Money.euros("2.00"));
        assertThat(summary.total()).isEqualTo(Money.euros("18.00"));
        assertThat(summary.appliedDiscounts()).hasSize(1);

        AppliedDiscount appliedDiscount = summary.appliedDiscounts().get(0);
        assertThat(appliedDiscount.promotionId()).isEqualTo(new PromotionId("PROMO-10-OFF-A"));
        assertThat(appliedDiscount.amount()).isEqualTo(Money.euros("2.00"));
        assertThat(appliedDiscount.target()).isEqualTo("line");
        assertThat(appliedDiscount.details()).isEqualTo("$2 off SKU-A");
    }

    @Test
    void shouldApplyMultiplePromotions() {
        Cart cart = new Cart(List.of(
            new LineItem(productA, new Quantity(2), Money.euros("10.00")),
            new LineItem(productB, new Quantity(1), Money.euros("25.00"))
        ));

        AppliedDiscount discount1 = new AppliedDiscount(
            new PromotionId("PROMO-10-OFF-A"),
            Money.euros("2.00"),
            "line",
            "10% off SKU-A"
        );

        AppliedDiscount discount2 = new AppliedDiscount(
            new PromotionId("PROMO-5-OFF-B"),
            Money.euros("5.00"),
            "line",
            "$5 off SKU-B"
        );

        StubPromotion promotion1 = new StubPromotion(
            new PromotionId("PROMO-10-OFF-A"),
            true,
            List.of(discount1)
        );

        StubPromotion promotion2 = new StubPromotion(
            new PromotionId("PROMO-5-OFF-B"),
            true,
            List.of(discount2)
        );

        PromotionEngine engine = new PromotionEngine(List.of(promotion1, promotion2));
        PriceSummary summary = engine.price(cart, context);

        // Subtotal: 2*$10 + 1*$25 = $45
        // Discounts: $2 + $5 = $7
        // Total: $45 - $7 = $38
        assertThat(summary.subtotal()).isEqualTo(Money.euros("45.00"));
        assertThat(summary.discountTotal()).isEqualTo(Money.euros("7.00"));
        assertThat(summary.total()).isEqualTo(Money.euros("38.00"));
        assertThat(summary.appliedDiscounts()).hasSize(2);
    }

    @Test
    void shouldNotApplyNonApplicablePromotions() {
        Cart cart = new Cart(List.of(
            new LineItem(productA, new Quantity(2), Money.euros("10.00"))
        ));

        StubPromotion nonApplicablePromotion = new StubPromotion(
            new PromotionId("PROMO-NOT-APPLICABLE"),
            false,  // Not applicable
            List.of()
        );

        PromotionEngine engine = new PromotionEngine(List.of(nonApplicablePromotion));
        PriceSummary summary = engine.price(cart, context);

        assertThat(summary.subtotal()).isEqualTo(Money.euros("20.00"));
        assertThat(summary.discountTotal()).isEqualTo(Money.euros("0.00"));
        assertThat(summary.total()).isEqualTo(Money.euros("20.00"));
        assertThat(summary.appliedDiscounts()).isEmpty();
    }

    @Test
    void shouldPreserveExplainabilityWithAllocations() {
        Cart cart = new Cart(List.of(
            new LineItem(productA, new Quantity(2), Money.euros("10.00")),
            new LineItem(productB, new Quantity(1), Money.euros("25.00"))
        ));

        Map<Sku, Money> allocations = Map.of(
            new Sku("SKU-A"), Money.euros("1.50"),
            new Sku("SKU-B"), Money.euros("2.50")
        );

        AppliedDiscount discount = new AppliedDiscount(
            new PromotionId("PROMO-MULTI-PRODUCT"),
            Money.euros("4.00"),
            "cart",
            "Multi-product discount",
            allocations
        );

        StubPromotion promotion = new StubPromotion(
            new PromotionId("PROMO-MULTI-PRODUCT"),
            true,
            List.of(discount)
        );

        PromotionEngine engine = new PromotionEngine(List.of(promotion));
        PriceSummary summary = engine.price(cart, context);

        assertThat(summary.subtotal()).isEqualTo(Money.euros("45.00"));
        assertThat(summary.discountTotal()).isEqualTo(Money.euros("4.00"));
        assertThat(summary.total()).isEqualTo(Money.euros("41.00"));

        AppliedDiscount appliedDiscount = summary.appliedDiscounts().get(0);
        assertThat(appliedDiscount.allocations()).hasSize(2);
        assertThat(appliedDiscount.allocations().get(new Sku("SKU-A"))).isEqualTo(Money.euros("1.50"));
        assertThat(appliedDiscount.allocations().get(new Sku("SKU-B"))).isEqualTo(Money.euros("2.50"));
    }
}
